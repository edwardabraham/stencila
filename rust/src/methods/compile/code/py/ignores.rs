use maplit::hashset;
use once_cell::sync::Lazy;
use std::collections::HashSet;

/// A list of Python symbols to ignore
///
/// This list was generated using
///    print(",\n".join([f"\"{symbol}\"" for symbol in dir(__builtins__)]))
/// inside Python 3.8.10 (default, Jun  2 2021, 10:49:15).
pub static USE_IGNORE: Lazy<HashSet<&str>> = Lazy::new(|| {
    hashset! {
        "ArithmeticError",
        "AssertionError",
        "AttributeError",
        "BaseException",
        "BlockingIOError",
        "BrokenPipeError",
        "BufferError",
        "BytesWarning",
        "ChildProcessError",
        "ConnectionAbortedError",
        "ConnectionError",
        "ConnectionRefusedError",
        "ConnectionResetError",
        "DeprecationWarning",
        "EOFError",
        "Ellipsis",
        "EnvironmentError",
        "Exception",
        "False",
        "FileExistsError",
        "FileNotFoundError",
        "FloatingPointError",
        "FutureWarning",
        "GeneratorExit",
        "IOError",
        "ImportError",
        "ImportWarning",
        "IndentationError",
        "IndexError",
        "InterruptedError",
        "IsADirectoryError",
        "KeyError",
        "KeyboardInterrupt",
        "LookupError",
        "MemoryError",
        "ModuleNotFoundError",
        "NameError",
        "None",
        "NotADirectoryError",
        "NotImplemented",
        "NotImplementedError",
        "OSError",
        "OverflowError",
        "PendingDeprecationWarning",
        "PermissionError",
        "ProcessLookupError",
        "RecursionError",
        "ReferenceError",
        "ResourceWarning",
        "RuntimeError",
        "RuntimeWarning",
        "StopAsyncIteration",
        "StopIteration",
        "SyntaxError",
        "SyntaxWarning",
        "SystemError",
        "SystemExit",
        "TabError",
        "TimeoutError",
        "True",
        "TypeError",
        "UnboundLocalError",
        "UnicodeDecodeError",
        "UnicodeEncodeError",
        "UnicodeError",
        "UnicodeTranslateError",
        "UnicodeWarning",
        "UserWarning",
        "ValueError",
        "Warning",
        "ZeroDivisionError",
        "_",
        "__build_class__",
        "__debug__",
        "__doc__",
        "__import__",
        "__loader__",
        "__name__",
        "__package__",
        "__spec__",
        "abs",
        "all",
        "any",
        "ascii",
        "bin",
        "bool",
        "breakpoint",
        "bytearray",
        "bytes",
        "callable",
        "chr",
        "classmethod",
        "compile",
        "complex",
        "copyright",
        "credits",
        "delattr",
        "dict",
        "dir",
        "divmod",
        "enumerate",
        "eval",
        "exec",
        "exit",
        "filter",
        "float",
        "format",
        "frozenset",
        "getattr",
        "globals",
        "hasattr",
        "hash",
        "help",
        "hex",
        "id",
        "input",
        "int",
        "isinstance",
        "issubclass",
        "iter",
        "len",
        "license",
        "list",
        "locals",
        "map",
        "max",
        "memoryview",
        "min",
        "next",
        "object",
        "oct",
        "open",
        "ord",
        "pow",
        "print",
        "property",
        "quit",
        "range",
        "repr",
        "reversed",
        "round",
        "set",
        "setattr",
        "slice",
        "sorted",
        "staticmethod",
        "str",
        "sum",
        "super",
        "tuple",
        "type",
        "vars",
        "zip"
    }
});
