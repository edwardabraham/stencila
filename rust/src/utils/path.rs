use lexiclean::Lexiclean;
use std::path::{Path, PathBuf};

/// Merge paths
///
/// Differs from `path.join()` in that it joins the *parent* of `path1`, not `path1` itself.
/// Does not canonicalize the path, or check for its existence.
pub fn merge<P1: AsRef<Path>, P2: AsRef<Path>>(path1: P1, path2: P2) -> PathBuf {
    path1.as_ref().join("..").join(path2).lexiclean()
}

#[cfg(test)]
mod test {
    use super::*;
    use path_slash::PathBufExt;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_merge() {
        assert_eq!(merge("a", "c").to_slash_lossy(), "c");
        assert_eq!(merge("a/b", "c").to_slash_lossy(), "a/c");
        assert_eq!(merge("a/b/../d/e", "c").to_slash_lossy(), "a/d/c")
    }
}
