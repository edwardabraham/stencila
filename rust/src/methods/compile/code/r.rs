use std::path::Path;

use super::{captures_as_args_map, child_text, is_quoted, remove_quotes, Compiler};
use crate::{
    graphs::{resources, Relation, Resource},
    utils::path::merge,
};
use itertools::Itertools;
use once_cell::sync::Lazy;

mod ignores;
use ignores::USES_IGNORE_FUNCTIONS;

/// Compiler for R
static COMPILER: Lazy<Compiler> = Lazy::new(|| {
    Compiler::new(
        tree_sitter_r::language(),
        r#"
(call
    function: (identifier) @function (#match? @function "^library|require$")
    arguments:(
        arguments
            ([(identifier)(string)] @arg)*
            (
                (identifier) @arg_name
                [(true)(false)] @arg_value
            )*
    )
)

(call
    function: (identifier) @function (#match? @function "^read\.")
    arguments: [
        (
            arguments
                .
                value: (string) @arg
        )
        (
            arguments
                name: (identifier) @arg_name
                .
                value: (string) @arg_value
        )
    ]
)

(call
    function: (identifier) @function (#match? @function "^write\.")
    arguments: [
        (
            arguments
                .
                value: (_) @arg
                .
                value: (string) @arg
        )
        (
            arguments
                name: (identifier) @arg_name
                .
                value: (string) @arg_value
        )
    ]
)

(program [
    (left_assignment name: (identifier) @identifer value: (_) @value)
    (equals_assignment name: (identifier) @identifer value: (_) @value)
])
(super_assignment name: (identifier) @identifer  value: (_) @value)

((identifier) @identifer)

"#,
    )
});

/// Compile some R code
pub fn compile(path: &Path, code: &str) -> Vec<(Relation, Resource)> {
    let code = code.as_bytes();
    let tree = COMPILER.parse(code);
    let captures = COMPILER.query(code, &tree);

    captures
        .into_iter()
        .filter_map(|(pattern, captures)| match pattern {
            0 => {
                // Imports a package using `library` or `require`
                let args = captures_as_args_map(captures);
                args.get("0")
                    .or_else(|| args.get("package"))
                    .and_then(|package| {
                        if let Some(is_char) = args.get("character.only") {
                            if is_char.starts_with('T') && !is_quoted(package) {
                                return None;
                            }
                        } else if is_quoted(package) {
                            return None;
                        }
                        Some((
                            Relation::Use,
                            resources::module("r", &remove_quotes(package)),
                        ))
                    })
            }
            1 => {
                // Reads a file
                let args = captures_as_args_map(captures);
                args.get("0").or_else(|| args.get("file")).map(|file| {
                    (
                        Relation::Read,
                        resources::file(&merge(path, remove_quotes(file))),
                    )
                })
            }
            2 => {
                // Writes a file
                let args = captures_as_args_map(captures);
                args.get("1").or_else(|| args.get("file")).map(|file| {
                    (
                        Relation::Write,
                        resources::file(&merge(path, remove_quotes(file))),
                    )
                })
            }
            3 | 4 => {
                // Assigns a symbol at the program root
                let name = captures[0].text.clone();
                let value = captures[1].node;
                let kind = match value.kind() {
                    // Attempt to infer the type of symbol
                    "true" | "false" => "Boolean",
                    "integer" => "Integer",
                    "float" => "Number",
                    "string" => "String",
                    "function_definition" => "Function",
                    "call" => match child_text(value, "function", code) {
                        "data.table" | "read.csv" | "read.csv2" | "read.delim" | "read.table" => {
                            "Datatable"
                        }
                        _ => "",
                    },
                    _ => "",
                };
                Some((Relation::Assign, resources::symbol(path, &name, kind)))
            }
            5 => {
                // Uses a function or variable
                let node = captures[0].node;
                let symbol = captures[0].text.clone();

                let mut parent = node.parent();
                while let Some(parent_node) = parent {
                    match parent_node.kind() {
                        // Skip identifiers that are the `name` of an assignment
                        "left_assignment" | "equals_assignment" | "super_assignment" => {
                            if let Some(name) = parent_node.child_by_field_name("name") {
                                if name == node {
                                    return None;
                                }
                            }
                        }
                        // Skip identifiers that are the `name` of a function call argument
                        "arguments" => {
                            let mut cursor = node.walk();
                            for name in parent_node.children_by_field_name("name", &mut cursor) {
                                if name == node {
                                    return None;
                                }
                            }
                        }
                        // Skip identifiers that are the `name` of a for loop, or that refer to it
                        "for" => {
                            if let Some(name) = parent_node.child_by_field_name("name") {
                                if name == node || name.utf8_text(code).unwrap() == symbol {
                                    return None;
                                }
                            }
                        }
                        // Skip package identifiers
                        "call" => {
                            let name = child_text(parent_node, "function", code);
                            if name == "library" || name == "require" {
                                return None;
                            }
                        }
                        // Skip identifiers within a function definition
                        "function_definition" => return None,
                        _ => {}
                    }
                    parent = parent_node.parent();
                }

                let resource = match node.parent() {
                    Some(parent_node) => match parent_node.kind() {
                        "call" => {
                            if USES_IGNORE_FUNCTIONS.contains(&symbol.as_str()) {
                                return None;
                            }
                            resources::symbol(path, &symbol, "Function")
                        }
                        _ => resources::symbol(path, &symbol, ""),
                    },
                    None => resources::symbol(path, &symbol, ""),
                };

                Some((Relation::Use, resource))
            }
            _ => None,
        })
        .unique()
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::tests::snapshot_content;
    use insta::assert_json_snapshot;
    use std::path::PathBuf;

    #[test]
    fn r_fragments() {
        snapshot_content("fragments/r/*.R", |path, code| {
            assert_json_snapshot!(compile(&PathBuf::from(path), code));
        });
    }
}
