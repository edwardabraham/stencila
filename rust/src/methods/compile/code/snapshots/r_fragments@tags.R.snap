---
source: rust/src/methods/compile/code/r.rs
expression: "compile(&PathBuf::from(path), code)"
input_file: fixtures/fragments/r/tags.R

---
[
  [
    {
      "use": [
        0,
        0,
        0,
        29
      ]
    },
    {
      "type": "Module",
      "language": "r",
      "name": "pkg1"
    }
  ],
  [
    {
      "use": [
        0,
        0,
        0,
        29
      ]
    },
    {
      "type": "Module",
      "language": "r",
      "name": "pkg2"
    }
  ],
  [
    {
      "use": [
        0,
        0,
        0,
        29
      ]
    },
    {
      "type": "Module",
      "language": "r",
      "name": "pk3"
    }
  ],
  [
    {
      "use": [
        0,
        0,
        0,
        29
      ]
    },
    {
      "type": "Module",
      "language": "r",
      "name": "pkg4"
    }
  ],
  [
    {
      "assign": [
        1,
        0,
        1,
        19
      ]
    },
    {
      "type": "Symbol",
      "path": "fragments/r/tags.R",
      "name": "var1",
      "kind": ""
    }
  ],
  [
    {
      "use": [
        2,
        0,
        2,
        12
      ]
    },
    {
      "type": "Symbol",
      "path": "fragments/r/tags.R",
      "name": "var1",
      "kind": ""
    }
  ],
  [
    {
      "read": [
        3,
        0,
        3,
        21
      ]
    },
    {
      "type": "File",
      "path": "file1"
    }
  ],
  [
    {
      "read": [
        3,
        0,
        3,
        21
      ]
    },
    {
      "type": "File",
      "path": "file2"
    }
  ],
  [
    {
      "write": [
        4,
        0,
        4,
        20
      ]
    },
    {
      "type": "File",
      "path": "file3"
    }
  ]
]
